### DISCUSSION TITLE
libpq broken on new postgres installations

### DISCUSSION THREAD

**Discussion by dvarrazzo** [Homebrew/discussions#3595] [Everyday usage]
Hello,

I am not a macOS user, so I can't provide the info required to file a bug.

New brew installation broke PostgreSQL client. See https://github.com/psycopg/psycopg2/issues/1490 for troubles with psycopg2, see failures in psycopg 3 test suite [in this run](https://github.com/psycopg/psycopg/runs/8133125829). More details will be tracked in https://github.com/psycopg/psycopg/issues/364.

As far as I can see, the libpq is now installed in a place where dynamic loading cannot find it. We are expecting this command to find the library:

```
python3 -c "import ctypes.util; print(ctypes.util.find_library('libpq.dylib'))"
```

please let us know what is needed to make this work again, thank you.

Related: #3579, #3584.

**Comment by carlocab** [comment]
How do you handle it for Homebrew on M1 macOS, which is installed in `/opt/homebrew`?

**Reply by dvarrazzo** [reply]
Whatever was before, it was working for M1 too. The [build/test script for M1](https://github.com/psycopg/psycopg/blob/master/tools/build/build_macos_arm64.sh) run successfully a few weeks ago.

**Reply by carlocab** [reply]
That doesn't tell me how you handle it, though, since the same issue should've been the case about M1 macOS previously as well. This is because `/opt/homebrew` is not in the dynamic loader's search path, so the command
```
python3 -c "import ctypes.util; print(ctypes.util.find_library('libpq.dylib'))"
```
wouldn't have worked on M1 macOS before the change in the directory structure of `postgresql`. (That is, unless you customise your Python build, but I don't think you do that.)

**Reply by dvarrazzo** [reply]
Maybe `eval "$(/opt/homebrew/bin/brew shellenv)"` does it?

**Reply by carlocab** [reply]
I don't think so, since all that does is `eval`s the following snippet:
```
❯ /opt/homebrew/bin/brew shellenv
export HOMEBREW_PREFIX="/opt/homebrew";
export HOMEBREW_CELLAR="/opt/homebrew/Cellar";
export HOMEBREW_REPOSITORY="/opt/homebrew";
export PATH="/opt/homebrew/bin:/opt/homebrew/sbin${PATH+:$PATH}";
export MANPATH="/opt/homebrew/share/man${MANPATH+:$MANPATH}:";
export INFOPATH="/opt/homebrew/share/info:${INFOPATH:-}";
```
As a workaround for now, I suggest doing
```
brew install libpq
brew unlink postgresql@14
brew link --force libpq
```
That should put a `libpq.dylib` in the same place you used to look.

Though it would be much better to understand why what you're doing works on M1 macOS, since that is probably the key to actually fixing this properly.

**Reply by dvarrazzo** [reply]
> Though it would be much better to understand why what you're doing works on M1 macOS, since that is probably the key to actually fixing this properly.

Note, if not clear: I don't know if this works on M1 **now**. I know it used to work a few weeks ago. The script I linked worked starting from a clean machine.

**Reply by carlocab** [reply]
> Note, if not clear: I don't know if this works on M1 **now**. I know it used to work a few weeks ago. The script I linked worked starting from a clean machine.

This was clear. I was, and am, interested in what makes M1 work before the change in the directory structure.

**Reply by dvarrazzo** [reply]
> As a workaround for now, I suggest doing
> 
> ```
> brew install libpq
> brew unlink postgresql@14
> brew link --force libpq
> ```
> 
> That should put a `libpq.dylib` in the same place you used to look.

Yes, but at the price of breaking Postgres altogether

https://github.com/psycopg/psycopg/runs/8175408336?check_suite_focus=true#step:10:229

Any timeline to fix this problem? This is a pretty serious breakage for everyone using postgres on macos.

**Reply by carlocab** [reply]
It's difficult to come up with a timeline until:

> > Note, if not clear: I don't know if this works on M1 **now**. I know it used to work a few weeks ago. The script I linked worked starting from a clean machine.
> 
> This was clear. I was, and am, interested in what makes M1 work before the change in the directory structure.

Without that then I have no idea why things are broken. It can't just be that the dynamic loader can't find `libpq`, because the dynamic loader doesn't check `/opt/homebrew` either and it seemed to work there.

**Reply by dvarrazzo** [reply]
I don't understand, homebrew people don't have a Mac M1 to run tests?

I can rent an M1 and provide you access if you give me an ssh public key. If we want to organize that (tomorrow, Europe timezone) please get in touch by email at daniele dot varrazzo at gmail.

**Reply by carlocab** [reply]
It's not about running tests, so access to an M1 machine is not the issue. There is no need to rent one. The issue is understanding your codebase and how you call `libpq`. I'll have a look.

Does psycopg work when installed with Anaconda?

**Reply by ani-ananth** [reply]
> > As a workaround for now, I suggest doing
> > ```
> > brew install libpq
> > brew unlink postgresql@14
> > brew link --force libpq
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That should put a `libpq.dylib` in the same place you used to look.
> 
> Yes, but at the price of breaking Postgres altogether
> 
> https://github.com/psycopg/psycopg/runs/8175408336?check_suite_focus=true#step:10:229
> 
> Any timeline to fix this problem? This is a pretty serious breakage for everyone using postgres on macos.

you don't need to break the postgres link (unless things were different in 2022), if you run:
```
brew link --force libpq --overwrite
```

only `clusterdb` has the conflict. and using the `overwrite` flag makes `libpq` get the `clusterdb` symlink preference over `postgresql`, while still having both brew links intact.

**Comment by carlocab** [comment]
Also, could you show the error you're currently getting with
```
python3 -c "import ctypes.util; print(ctypes.util.find_library('libpq.dylib'))"
```
? If you could show a backtrace that would also be very helpful.

**Reply by dvarrazzo** [reply]
This statement returns None. It doesn't fail, but doesn't find anything either.

**Comment by carlocab** [comment]
I suggest using `pg_config` to query the location of `libpq`. This works for me:
```python3
import ctypes
import subprocess

libpq_dir = subprocess.run(['pg_config', '--libdir'], capture_output=True, text=True).stdout.strip()
ctypes.cdll.LoadLibrary(f'{libpq_dir}/libpq.dylib')
```

**Reply by dvarrazzo** [reply]
The pg_config binary is not a requirement for psycopg and doesn't get necessary installed with the libpq. It is a development tool, not a runtime tool.

**Reply by carlocab** [reply]
Yes, but you can fall back to it if `find_library` fails. It is also included in the postgres installations you are having issues with.

`find_library` only works on Darwin if you install directly into `/usr/local` or if Python has been configured in a certain way. Neither are good assumptions to make.

**Comment by dvarrazzo** [comment]
Yes, I understand that using pg_config as a fallback is a possible way to get out of the current state. I'll try to make an emergency release with this hack.

**Comment by frou** [comment]
One more datapoint: If as a user, we run some Python process as a macOS launchd job/LaunchAgent, then `/usr/local/bin` is not in `PATH` by default and so `pg_config` cannot even be run.

**Reply by carlocab** [reply]
You can specify your `PATH` in the launch agent `plist`.

This was going to happen eventually anyway: Apple is no longer going to release new Intel Macs, and `/opt/homebrew/lib` is not in `dyld`'s default search path.

**Comment by salex** [comment]
I guess I ran into the same problem. I had not used brew in a while and wanted to upgrade git or something. Had to upload new Xcode command line tools and decided to do a brew upgrade/update.

That updated postgresql@13(installed) also installed postgresql@14.  I tried to follow all the warning - I've been through it before and remember clobbering my db. I then did a BoBo and 

```sh
mv -v "/opt/homebrew/var/postgres" "/opt/homebrew/var/postgresql@14"
```
which I think overwrote the @14 version. Think I changed to PG_VERSION to 14 and tried starting postgres. It started, no data.

I did a few hours and looking and reading. It installed @14, but found out if I did `psql postgres` I found my database in the @14 version.

Though I was saveed, but ran into the lib error when I tried to start rails (bin/dev)

```log 
SteveAir:our_books $ bin/dev
15:23:58 web.1  | started with pid 13129
15:23:58 css.1  | started with pid 13130
15:23:59 web.1  | /Users/salex/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/bootsnap-1.13.0/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:32:in `require': dlopen(/Users/salex/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/pg-1.4.2/lib/pg_ext.bundle, 0x0009): Library not loaded: '/opt/homebrew/opt/postgresql/lib/libpq.5.dylib' (LoadError)
15:23:59 css.1  | rails aborted!
15:23:59 web.1  |   Referenced from: '/Users/salex/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/pg-1.4.2/lib/pg_ext.bundle'
15:23:59 css.1  | LoadError: dlopen(/Users/salex/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/pg-1.4.2/lib/pg_ext.bundle, 0x0009): Library not loaded: '/opt/homebrew/opt/postgresql/lib/libpq.5.dylib'
15:23:59 web.1  |   Reason: tried: '/opt/homebrew/opt/postgresql/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file), '/opt/homebrew/Cellar/postgresql@14/14.5_3/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file) - /Users/salex/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/pg-1.4.2/lib/pg_ext.bundle
15:23:59 css.1  |   Referenced from: '/Users/salex/.rbenv/versions/3.0.1/lib/ruby/gems/3.0.0/gems/pg-1.4.2/lib/pg_ext.bundle'
15
...
```

So it looks like I'm stuck until a solution is found the the lib load!

Guess I'll take a few day off. or point me a way back to @13

**Reply by carlocab** [reply]
Try reinstalling the `pg_ext` gem.

**Reply by salex** [reply]
That broke everthing!

I assume you meant the pg gem, could not find a reference to pg_ext. It (pg) was a 1.4.2.  I had band-aided it yesterday by add a symbolic link in /usr/local/lib to ibpq.5.dylib. Everything was working. Then I tried the update suggested and that broke it and I can't get back to the hack.

```
our_books $ psql postgres
psql: error: connection to server on socket "/tmp/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
```

the bundle update did install rails 7.0.4. I may try to move back to 7.0.3 later today.

**Reply by salex** [reply]
After some time trying to figure out the brew postgres symbolic linking, my initial 

```
mv -v "/opt/homebrew/var/postgres" "/opt/homebrew/var/postgresql@14"
```

moved my @13 database to @14 which was named `postgres` . So my linking was all screwed up. I added a `postgres` link to @14 and got it working.  I still had the symbolic link in /usr/local/lib to ibpq.5.dylib (band-aid).  Removed it and guess the reinstall worked - no error.
